Fixes #10521 - reporting templates

ares opened the pull request:
![reporting](https://user-images.githubusercontent.com/109773/42346988-ab8d918a-80a4-11e8-9042-d9202c87aa0a.gif)

This is an initial version that does the rendering in web process. It comes with only one built-in report. It also provides API, hammer support will be trivial.

In future but not in this PR, I'd like to add 
1) template inputs similar to how it works with JobTemplates, that's currently blocked by https://github.com/theforeman/foreman/pull/5545
2) scheduling generating to future
3) offloading to dynflow executor
4) enrich the template DSL with more helpers for commonly asked reports, implement more reports
5) upload report templates to community templates repo (blocked by merging this PR)
6) verify foreman_templates works, should work out of the box though
7) write a chapter in manual (blocked by merging this PR)
8) add support to hammer

For those who discussed the original design, I came to conclusion that streaming the rendered template is not possible if we want to leverage power of erb for rendering itself. Loading of objects should be handled by new macros (see `load_hosts`) that ensures we process objects in batches. Rendering will be shortly possible asynchronously by dynflow, we'll probably deliver the result via email in such case, so for big reports, web process rendering/preview will be only useful for reasonable sizes.

<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` or `Refs #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Extract all strings for i18n, see [Translating section in the guide]
(https://projects.theforeman.org/projects/foreman/wiki/Translating)
* Prepend `[WIP]` for work in progress from bots triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* Be nice and respectful

We are running bots that will poke you if you miss an item from the list :-)

--->

ares committed:
Fixes #10521 - Adds renderable report templates

ares commented:
cc @tstrachota, please take a look when you have a chance, I'd appreciate help with steps 1. and 2. where I'd like to implement the scheduling form using react

timogoebel commented on app/controllers/report_templates_controller.rb:
```diff
@@ -0,0 +1,31 @@
+  end
+
+  def generate
+    find_resource
```
Nit: I'd use a before_action for this.

ares replied:
I wanted too, but it's inherited from templates_controller, if I specify it here, I'd need to copy all listed actions from templaters_controller, adding just generate action here overrides the whole list. I could extract the list of action to some constant/method but this felt slightly more straightforward.

timogoebel commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -0,0 +1,139 @@
+
+      private
+
+      def load_vars_from_report_template
```
Why is this needed?

ares replied:
leftover, removed

timogoebel commented on lib/foreman/renderer.rb:
```diff
@@ -236,21 +240,37 @@ def dns_lookup(name_or_ip)
       raise e
     end
 
+    def load_hosts(search: '', includes: nil)
```
I believe we should not add this here. Does it make sense to have this method e.g. in a provisioning template? This method will be available in the host object as well. `Host.first.load_hosts` does not make a lot of sense. Luckily #5683 will provide an easy way to make this possible.

ares replied:
I agree, once the refactoring is done, I'd create a new scope for reporting templates (or other ways around, based on what gets in first)

timogoebel reviewed (commented):
Not a full review, just some comments I had while reading the code.

timogoebel commented:
> offloading to dynflow executor

What would be gain by rendering the template in the dynflow executor?

tstrachota commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -0,0 +1,139 @@
+      api :GET, "/report_templates/", N_("List all report templates")
+      api :GET, "/locations/:location_id/report_templates/", N_("List all report templates per location")
+      api :GET, "/organizations/:organization_id/report_templates/", N_("List all report templates per organization")
+      param :operatingsystem_id, String, :desc => N_("ID of operating system")
```
Copied over by mistake?

tstrachota commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -0,0 +1,139 @@
+        process_response @report_template.save
+      end
+
+      api :POST, "/report_templates/import", N_("Import a provisioning template")
```
`"Import a report template"`

tstrachota commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -0,0 +1,139 @@
+        process_response @report_template.save
+      end
+
+      api :GET, '/report_templates/:id/export', N_('Export a partition template to ERB')
```
`s/partition template/report template`

tstrachota commented on app/views/report_templates/_diff.html.erb:
```diff
@@ -0,0 +1,6 @@
+<div class="editor-container">
+  <%= text_area_tag "diffLines","" ,:size => "12x30", :class => "col-md-12 editor_source diffMode", :label=>_("Template diff"), :'data-file-name' => file_name%>
```
Weird spacing:
` "diffLines","" ,:size `
vs.
` "diffLines", "", :size `
and also add space before `%>` please

tstrachota reviewed (changes requested):
I haven't done any deep testing yet. Sending just first batch of small issues I came across while reading the code.

ares commented:
fixed/replied to all comments I received so far, one question raised by @timogoebel 

> What would be gain by rendering the template in the dynflow executor?

Originally I thought we'd use the same workflow for "render now" and "render later" and potentially the web process wouldn't be blocked in case the report is rendered for a while, dynflow executor can run not only as a separate process but also on a separate host. I realized this would be quite complicated with ActiveJob today, so offloading will be possible only with delayed rendering. Luckily we can use ActiveJob with dynflow backed for this now.

tstrachota commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -0,0 +1,134 @@
+          when 'export'
+            'view'
+          when 'generate'
+            'generate_report_templates'
```
I guess you should use only 'generate' here.

ares replied:
fixed

tstrachota commented on app/controllers/report_templates_controller.rb:
```diff
@@ -0,0 +1,30 @@
+  def action_permission
+    case params[:action]
+      when 'generate'
+        'generate_report_templates'
```
Same here, 'generate' is the right value.

ares replied:
fixed

tstrachota reviewed (changes requested):
Thanks for your updates @ares. The code looks good to me.

I tested the functionality and I have troubles with rendering reports as a regular user (not an admin). I always get 404 `Resource report_template not found by id '110'` even though I can view and edit the template normally. I have the "generate" permission too. 

After some investigation I found out that the permission created in `scope_for` is `"generate_report_templates_report_templates"`. Both UI and API controllers are affected.

tstrachota commented:
@ares I'm little concerned about the preview functionality for large reports. If I understand it correctly it triggers rendering on each 'Preview' click. Won't it be too much of unexpected burden?

ares commented:
> @ares I'm little concerned about the preview functionality for large reports. If I understand it correctly it triggers rendering on each 'Preview' click. Won't it be too much of unexpected burden?

I agree with the concern, I'm waiting on templates refactoring to be merged to introduce also rendering mode (real/preview). In preview mode we could limit all resource loader with `.limit(10)` for example.

ares commented:
Thanks for thorough testing under non-admin user. I pushed the permission change into last commit, please take another look.

ares commented:
rebased after medium rendering was merged

tstrachota commented:
Thanks @ares! Some tests are failing. Apart from that it's good to go.

ares commented:
2 failures should be fixed in last commit, 3rd is unrelated and this should fix it https://github.com/theforeman/foreman/pull/5882

ares commented:
rebased to make Foreman tests green, squashed as asked

tstrachota commented:
Thanks, waiting for results of the tests to merge.

timogoebel commented:
> Thanks, waiting for results of the tests to merge.

@ares said he already has a branch that rebased on the template scope PR. Let's wait for the other PR to get in before we merge this. I can hit the button next week, no problem.

timogoebel commented on lib/foreman/renderer/scope/macros/base.rb:
```diff
@@ -78,11 +78,41 @@ def default_template_url(template, hostgroup)
                     :host => host, :port => port)
           end
 
+          def load_hosts(search: '', includes: nil)
```
Do we want to have this in the base scope and available in all templates? I think it makes sense to move this to a dedicated scope for reports.

ares replied:
I thought this would be useful everwhere, e.g. in host's kickstart template, you may want to render a config file based on data of other hosts, e.g. setting up a gateway, /etc/hosts etc

timogoebel commented on test/unit/foreman/renderer/renderers_shared_tests.rb:
```diff
@@ -137,5 +137,15 @@ module RenderersSharedTests
       end
     end
+
+    test "should load hosts" do
```
I'd move these tests to a dedicated scope class for report templates.

ares replied:
same as the previous one

timogoebel reviewed (changes requested):
@ares: Thanks, I have two more suggestions after the rebase.

timogoebel commented on app/controllers/templates_controller.rb:
```diff
@@ -109,8 +109,10 @@ def resource_name
   private
 
-  def safe_render(template)
-    render :plain => @host.render_template(template: template, params: params)
```
😢 This method had a short life :-(
We should ideally keep this as unattended controller still uses the method.

timogoebel replied:
In addition, the reporting template should probably use it's own scope.
Maybe we need to use `template.get_scope` or something...

ares replied:
my exact thoughts after our IRC discussion, see the latest commit, I couldn't resist :-) we can keep it for now or I can do the similar change there if you prefer

timogoebel replied:
I like the default class defined for every template type. 

How about we introduce a `template.render` method that looks similar to this:

```
def render(template:, mode: Foreman::Renderer::REAL_MODE, host: nil, params: {})
  source = Foreman::Renderer.get_source(template: self, host: host)
  scope = Foreman::Renderer.get_scope(host: host, params: params, mode: mode, template: self)
  Foreman::Renderer.render(source, scope)
end
```

That could replace host.render_template but we dry this up a lot more.

ares committed:
Refs #10521 - rebase also API controller

ares committed:
Refs #10521 - use separate report scope

tstrachota commented:
It works good and the design seems fine to me either, but I'll leave the final word on @timogoebel since he has a lot more insights into the latest template refactoring.

timogoebel commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -0,0 +1,136 @@
+    class ReportTemplatesController < V2::BaseController
+      include Foreman::Controller::Parameters::ReportTemplate
+      include Foreman::Controller::TemplateImport
+      include UnattendedHelper
```
Luckily, we don't need this anymore.

timogoebel commented on app/controllers/api/v2/report_templates_controller.rb:
```diff
@@ -0,0 +1,136 @@
+      include Foreman::Controller::Parameters::ReportTemplate
+      include Foreman::Controller::TemplateImport
+      include UnattendedHelper
+      include Foreman::Controller::ProvisioningTemplates
```
Why do we need this? `load_vars_from_template` is not used here anywhere.

ares replied:
no longer needed, I know that without this, renderer used to fail since it tried to call `load_vars_from_template`, which is why we also had this method empty here, it works fine after refactoring

timogoebel commented on lib/foreman/renderer/base_renderer.rb:
```diff
@@ -16,11 +16,13 @@ def render
-          template_name: source.name,
-          template_host_name: scope.host.try(:name),
-          template_host_id: scope.host.try(:id))
+        if !scope.preview? && source.template.is_a?(ReportTemplate)
```
I might be a little overly strict here, but is there a way we can use inheritance for this?
Maybe `if !scope.previw && template.log_results?` ?

timogoebel reviewed (changes requested):
@ares: I went over the code one more time with more focus. Left some comments.
In general: Nice, we're close.

ares committed:
Refs #10521 - fix comments

ares commented:
I hope I addressed all comments, I like `template.render` idea, it deduplicated the code a lot. Also we no longer need provisioning_template render_template method.

timogoebel reviewed (approved):
LGTM, APJ.

timogoebel commented on app/controllers/templates_controller.rb:
```diff
@@ -109,8 +109,8 @@ def resource_name
-  def safe_render(template)
-    render :plain => @host.render_template(template: template, params: params)
+  def safe_render(template, mode = Foreman::Renderer::REAL_MODE)
+    render :plain => template.render(host: @host, params: params, mode: mode, template: template)
```
@ares: Can you please also change this in the unattended controller to be consistent? Thanks.

ares replied:
pushed in separate commit, good point

timogoebel reviewed (changes requested):
I found one small nit, otherwise still 👍.

kamils-iRonin commented on lib/foreman/renderer.rb:
```diff
@@ -28,8 +32,9 @@ def get_source(klass: Foreman::Renderer::Source::Database, template:, **args)
 
-      def get_scope(klass: Foreman::Renderer::Scope::Provisioning, host: nil, params: {}, variables: {})
-        klass.new(host: host, params: params, variables: variables)
+      def get_scope(klass: nil, host: nil, params: {}, variables: {}, mode: REAL_MODE, template: nil)
```
Maybe we can pass `klass` here, instead of `template` instance?

ares replied:
sorry, not sure I follow, if we pass klass as an argument, it wins `klass ||= `, then we default to template's default class and then we fallback to Foreman::Renderer::Scope::Provisioning

ares committed:
Refs #10521 - fix tests

ares committed:
Refs #10521 - update unattended controller

ares commented:
@timogoebel tests are green :-)

timogoebel commented:
Thanks, @ares.
